cumulus/client/cirrus-executor/src/bundle_processor.rs

```diff
@@ -285,64 +282,68 @@ where
-						"Generating the signed execution receipt for #{}",
-						header_hash
-					);
+		let best_execution_chain_number =
+			<NumberFor<Block>>::decode(&mut best_execution_chain_number.encode().as_slice())
+				.expect("Primary number and secondary number must use the same type; qed");
 
-					let signed_execution_receipt = SignedExecutionReceipt {
-						execution_receipt,
-						signature: ExecutorSignature::decode(&mut signature.as_slice()).map_err(
-							|err| {
-								sp_blockchain::Error::Application(Box::from(format!(
-									"Failed to decode the signature of execution receipt: {err}"
-								)))
-							},
-						)?,
-						signer: executor_id,
-					};
-
-					if let Err(e) = self
-						.execution_receipt_sender
-						.unbounded_send(signed_execution_receipt.clone())
-					{
-						tracing::error!(target: LOG_TARGET, error = ?e, "Failed to send signed execution receipt");
-					}
+		assert!(
+			header_number > best_execution_chain_number,
+			"Consensus chain number must larger than execution chain number by at least 1"
+		);
 
-					// Return `Some(_)` to broadcast ER to all farmers via unsigned extrinsic.
-					Ok(Some(signed_execution_receipt))
-				},
-				Ok(None) => Err(sp_blockchain::Error::Application(Box::from(
-					"This should not happen as the existence of key was just checked",
-				))),
-				Err(error) => Err(sp_blockchain::Error::Application(Box::from(format!(
-					"Error occurred when signing the execution receipt: {error}"
-				)))),
-			}
+		// Ideally, the receipt of current block will be included in the next block, i.e., no
+		// missing receipts.
+		if header_number == best_execution_chain_number + One::one() {
+			self.try_sign_and_send_receipt(primary_hash, execution_receipt)
 		} else {
-			Ok(None)
+			// Receipts for some previous blocks are missing.
+			let max_drift = self
+				.primary_chain_client
+				.runtime_api()
+				.maximum_receipt_drift(&BlockId::Hash(primary_hash))?;
+
+			let max_drift = <NumberFor<Block>>::decode(&mut max_drift.encode().as_slice())
+				.expect("Primary number and secondary number must use the same type; qed");
+
+			let max_allowed = (best_execution_chain_number + max_drift).min(header_number);
+
+			// TODO: parallelize and avoid spamming the missing receipts?
+			let mut to_send = best_execution_chain_number;
+			while to_send <= max_allowed {
+				let block_hash = self.client.hash(to_send)?.ok_or_else(|| {
+					sp_blockchain::Error::Backend(format!("Hash for Block {:?} not found", to_send))
+				})?;
+
+				match crate::aux_schema::load_execution_receipt(&*self.client, block_hash)? {
+					Some(receipt) => {
+						self.try_sign_and_send_receipt(primary_hash, receipt)?;
+					},
+					None => {
+						// TODO: In order to solve the problem that the receipt can be pruned by
+						// executor, we need to check if every ER in the primary chain is valid beforehand:
+						// - If ER is invalid, cache it and expect FraudProof within next X blocks.
```

Comment:
Why waiting for someone else to submit a FP though?

Reply:
Actually, it's not to wait for someone else to submit a FP, the context for this line is that the major sync is still ongoing, you just check if a FP can be found in the next X blocks after an invalid receipt was found.

Reply:
But this piece of code doesn't seem to know whether major sync is ongoing or not. So it could as well send FP and get it invalidated by execution chain if it happens to be unnecessary, no?

Reply:
Only when the major sync is finished we'll start to send FP for the first incorrect ER, otherwise, we just cache the incorrect ER and see if the FR can be found in the next X blocks.

Reply:
That is what I'm saying: how does this code know when major sync is over?

Reply:
We already have this check? https://github.com/subspace/subspace/blob/953056c2076a042536e6f3f5a9f339f0292631c4/cumulus/client/cirrus-executor/src/bundle_processor.rs#L280